sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamAggregationRule.java

```diff
@@ -54,29 +62,133 @@ public void onMatch(RelOptRuleCall call) {
 
   private static RelNode updateWindow(RelOptRuleCall call, Aggregate aggregate, Project project) {
     ImmutableBitSet groupByFields = aggregate.getGroupSet();
-    List<RexNode> projectMapping = project.getProjects();
+    ArrayList<RexNode> projects = new ArrayList(project.getProjects());
 
-    Optional<AggregateWindowField> windowField = Optional.empty();
+    WindowFn windowFn = null;
+    int windowFieldIndex = -1;
 
     for (int groupFieldIndex : groupByFields.asList()) {
-      RexNode projNode = projectMapping.get(groupFieldIndex);
+      RexNode projNode = projects.get(groupFieldIndex);
       if (!(projNode instanceof RexCall)) {
         continue;
       }
 
-      windowField = AggregateWindowFactory.getWindowFieldAt((RexCall) projNode, groupFieldIndex);
+      RexCall rexCall = (RexCall) projNode;
+      WindowFn fn = createWindowFn(rexCall.getOperands(), rexCall.op.kind);
+      if (fn != null) {
```

Comment:
nit: this is exactly what `Optional` is for

Reply:
My understanding is that `Optional` is for passing null through lambda expressions and maps. In this simple case it makes things more complicated.